packages/xo-server/src/xo-mixins/backups-ng-logs.js

```diff
@@ -65,139 +71,208 @@ const taskTimeComparator = ({ start: s1, end: e1 }, { start: s2, end: e2 }) => {
-      const [jobLogs, restoreLogs, restoreMetadataLogs] = await Promise.all([
-        this.getLogs('jobs'),
-        this.getLogs('restore'),
-        this.getLogs('metadataRestore'),
-      ])
-
-      const { runningJobs, runningRestores, runningMetadataRestores } = this
-      const consolidated = {}
-      const started = {}
-
-      const handleLog = ({ data, time, message }, id) => {
-        const { event } = data
-        if (event === 'job.start') {
-          if (
-            (data.type === 'backup' || data.key === undefined) &&
-            (runId === undefined || runId === id)
-          ) {
-            const { scheduleId, jobId } = data
-            consolidated[id] = started[id] = {
-              data: data.data,
-              id,
-              jobId,
-              jobName: data.jobName,
-              message: 'backup',
-              scheduleId,
-              start: time,
-              status: runningJobs[jobId] === id ? 'pending' : 'interrupted',
-            }
-          }
-        } else if (event === 'job.end') {
-          const { runJobId } = data
-          const log = started[runJobId]
-          if (log !== undefined) {
-            delete started[runJobId]
-            log.end = time
-            log.status = computeStatusAndSortTasks(
-              getStatus((log.result = data.error)),
-              log.tasks
-            )
-          }
-        } else if (event === 'task.start') {
-          const task = {
+export default class BackupNgLogs {
+  constructor(app, { backup }) {
+    this._app = app
+    this.getBackupNgLogs = debounceWithKey(
+      this.getBackupNgLogs,
+      10e3,
+      runId => runId
+    )
+
+    app.on('clean', () =>
+      app.getStore(STORE_NAMESPACE).then(db => db.gc(backup.nKeptTasks))
+    )
+  }
+
+  async getBackupNgLogs(runId?: string) {
+    const app = this._app
+    const backupTaskStore = await app.getStore(STORE_NAMESPACE)
+
+    const [
+      jobLogs,
+      restoreLogs,
+      restoreMetadataLogs,
+      backupTasks,
+    ] = await Promise.all([
+      app.getLogs('jobs'),
+      app.getLogs('restore'),
+      app.getLogs('metadataRestore'),
+      backupTaskStore.getAll(),
+    ])
+
+    if (runId !== undefined && backupTasks[runId] !== undefined) {
+      return backupTasks[runId]
+    }
+
+    const { runningJobs, runningRestores, runningMetadataRestores } = app
+    const consolidated = {}
+    const started = {}
+
+    // used in order to clean subtasks when the global task finish
+    const tasksByTopParent = {}
+    // an optimization to get the hight level parent
+    const taskWithTopParent = {}
+
+    const finishedTasks = []
+    const storeBackupTasks = async () => {
+      const logsStore = await app.getStore('logs')
+      return asyncMap(finishedTasks, async id => {
+        await backupTaskStore.put(id, consolidated[id])
+        return asyncMap(tasksByTopParent[id], id => logsStore.del(id))
+      })
+    }
```

Comment:
Why this function instead of inline code?

Reply:
It's for readability and because we wan't wait the storing of the consolidated logs and in the storing implementation, there have some tasks which should be before others.

Reply:
I'm sorry, I don't understand, we'll have to discuss about this directly 🙂